[WIP] Structured parameters

hadware opened the pull request:
This adds supports for "dynamically defined" structured parameters for pipelines:

```python
from pyannote.pipeline import Pipeline
from pyannote.pipeline.parameter import Uniform, Integer, ParamDict, ParamList, Categorical


class MyPipeline(Pipeline):

    def __init__(self):
        super().__init__()

        self.mydict = ParamDict({
            "param_a": Uniform(0, 1),
            "param_b": Integer(0, 10)
        })
        
        self.mylist = ParamList([
            Integer(0, i) for i in range(5)
        ])
        
        # structured parameters can also be nested (list of dicts)
        self.nested = ParamList([
            ParamDict({
                "param_a": Uniform(0, 1),
                "param_b": Integer(0, 10)
            }),
            Categorical(["a", "b", "c"]),
            Uniform(0, 100)
        ])
 ```

This is a work in progress, i'm going to add tests in the near future, although i'd be willing to have your feedback on the implementation. I tried integrating it as well as i could with the current state of things, please don't hesitate on telling me if something might break because of this.

I didn't properly comment/docstring everything for now, i'm waiting for my tests to work and your approval before doing so.

hbredin commented on pyannote/pipeline/parameter.py:
```diff
@@ -25,16 +25,164 @@
+        params_dict = {}
+        structured_params = {name: {} for name, param in self._params.items()
+                             if isinstance(param, StructuredParameter)}
+        for name, value in flattened_params.items():
+            tokens = name.split(">")
+            root_name: str = tokens[0]
+            if len(tokens) > 1 and root_name in structured_params:
+                subparam_name = ">".join(tokens[1:])
+                structured_params[root_name][subparam_name] = value
+            else:
+                params_dict[name] = value
```
Could please add some comments to clarify what is going on here?

hbredin commented on pyannote/pipeline/pipeline.py:
```diff
@@ -94,7 +94,8 @@ def __setattr__(self, name, value):
         If `value` is an instance of `Parameter`, store it in `_parameters`.
         If `value` is an instance of `Pipeline`, store it in `_pipelines`.
-        If `name` is in `_parameters`, store `value` in `_instantiated`.
+        If `name` is in `_parameters`, and `value` isn't an instance of `Parameter`
```
Maybe rewrite this as 
```
If `value` ...
elif `value`...
elif `name`...
```
This way, you don't have to clarify "`value` isn't an instance of Parameter"

hbredin commented on pyannote/pipeline/pipeline.py:
```diff
@@ -165,21 +166,30 @@ def _flattened_parameters(
             params = {
                 n: p.value for n, p in self._parameters.items() if isinstance(p, Frozen)
             }
+            params = dict(flatten_structured(params))
```
What if only part of the structured parameters are frozen? Does this work? 
Or do I understand this correctly that either the whole structured parameters is frozen or it's not.

hadware replied:
ah, this is tricky. In _theory_, you could very well freeze subparameters, although in this case, it'll consider only the frozen "root" parameter. This could be fixed, but the freezing logic probably should be overhauled as well. My take on this issue was that "people" are mostly going to freeze the whole structure when needed.

hadware replied:
An afterthought on this issue: what would be needed to have a working freezing capability even for structured parameters:
- make the `_flatten` method able to search the structured parameters for frozen subparameters
- add a `freeze` method to structured parameters, so users can do something like `my_pipeline.parameters().param_dict["param_a"].freeze(value)`

hbredin commented on tests/test_pipeline.py:
```diff
@@ -1,5 +1,70 @@
 
-def test_params():
-    assert 2 == 1 + 1
+def test_pipeline_params_simple():
```
Would be nice to also add tests that actually run the optimizer...

hadware replied:
I made some very rudimentary tests, if you have any suggestions i'd be glad to know

hbredin reviewed (commented):
Thanks for working on this.
Here are a bunch of comments based on a first pass I just did.

What if we added a `.items()` to any parameters (including non-structured ones)?

```python
for name, value in Uniform(...).items():
    pass    
```

Maybe it would simplify the code base as you wouldn't have to do:
```python
if isinstance(..., StructuredParam):
   ...
else:
  ...
```

hadware commented:
> Thanks for working on this.
> Here are a bunch of comments based on a first pass I just did.
> 
> What if we added a `.items()` to any parameters (including non-structured ones)?
> 
> ```python
> for name, value in Uniform(...).items():
>     pass    
> ```
> 
> Maybe it would simplify the code base as you wouldn't have to do:
> 
> ```python
> if isinstance(..., StructuredParam):
>    ...
> else:
>   ...
> ```

That is _exactly_ what I kind of hesitated on doing, although for a first "pass" on this issue I wanted to keep things simple and not over-engineer everything. If you're willing to go down that path, I'd even suggest making a mixin that could be shared among parameters and pipelines (since pipelines _kind of_ are structured parameters as well).

hbredin commented:
> That is _exactly_ what I kind of hesitated on doing, although for a first "pass" on this issue I wanted to keep things simple and not over-engineer everything. If you're willing to go down that path, I'd even suggest making a mixin that could be shared among parameters and pipelines (since pipelines _kind of are_ structured parameters as well).

_Les grands esprits..._

Not sure about using it for pipelines though...

Let's keep this PR focused on adding this new "structured parameters" feature.

We'll decide later if we factorize the code for use with pipelines.

hadware commented on pyannote/pipeline/parameter.py:
```diff
@@ -25,17 +25,126 @@
+        A list of parameters
+    """
+
+    def __init__(self, params: List[Parameter]):
```
Done

hbredin commented:
This `freezing` thing bothers me :-)

What if we did things completely differently and made `ParamDict` a subclass of `Pipeline`?

```python
from pyannote.pipeline import Pipeline
from pyannote.pipeline.parameter import Uniform

class ParamDict(Pipeline):    
    def __init__(self, **params):
        super().__init__()
        for param_name, param_value in params.items():
            setattr(self, param_name, param_value)
    
    def __getitem__(self, param_name):
        return getattr(self, param_name)
 
class DemoPipeline(Pipeline):
    def __init__(self):
        super().__init__()
        self.regular_param = Uniform(0., 1.)
        self.param_dict = ParamDict(probability=Uniform(0.0, 1.0), 
                                    angle=Uniform(0.0, 3.14))

pipeline = DemoPipeline()
pipeline.parameters()

pipeline.freeze({"param_dict": {"probability": 0.5}})
pipeline.parameters(frozen=True)
# {'param_dict': {'probability': 0.5}}

pipeline.param_dict["probability"]  # thanks to ParamDict.__getitem__
```

Not sure how to handle `ParamList` though.

hadware commented:
:thinking: If you think this really has t be fixed, i'm not a huge fan of making `ParamDict` inherit from `Pipeline`. If you're ok with that, i'd rather do the solution I gave there https://github.com/pyannote/pyannote-pipeline/pull/34#discussion_r716824172

hadware commented:
Alright, I fixed the freezing issue. The takeaway is that : 
- `ParamDict` can now be freezed partly (or entirely)
- `ParamList` can only be frozen entirely (else the indexing makes it a nightmare), but parts of a list of dicts (for instance) can be frozen still. the "workaround" to freeze parts of a list is to instead use a dict indexed by ints.
- I've added tests for that as well

hbredin commented on tests/test_pipeline.py:
```diff
@@ -1,5 +1,113 @@
+
+    pl = TestPipeline()
+    frozen_params = {'nested_param': [{'param_b': 7, 'param_c': {'param_d': 2}}, "a", 70]}
+    pl.instantiate(pl.parameters(FakeTrial()))
+    pl.freeze(frozen_params)
```
A real use case would be to freeze before instantiation.

hbredin commented on tests/test_optimizer.py:
```diff
@@ -0,0 +1,116 @@
+        def get_direction(self) -> Direction:
+            return direction
+
+    optimizer_tester(pipeline=SumPipeline(), target=target)
```
Can you also add an optimizer test with (partially) frozen pipelines?

hadware replied:
Done

hbredin commented on tests/utils.py:
```diff
@@ -0,0 +1,19 @@
+from optuna.distributions import CategoricalChoiceType
+
+
+class FakeTrial(Trial):
```
This `FakeTrial` is a brilliant idea!

hadware replied:
Thanks! It did take me a fair deal of digging in optuna's code to get that right tbh :stuck_out_tongue:

hadware commented on pyannote/pipeline/pipeline.py:
```diff
@@ -93,8 +93,9 @@ def __setattr__(self, name, value):
-        If `value` is an instance of `Pipeline`, store it in `_pipelines`.
-        If `name` is in `_parameters`, store `value` in `_instantiated`.
+        elif `value` is an instance of `Pipeline`, store it in `_pipelines`.
+        elif `value` isn't an instance of `Parameter` and `name` is in `_parameters`,
```
"we" do, but i wanted to be super clear about that (since, on my part, it took me a bit of guessing to be sure of what it id).

hbredin commented on pyannote/pipeline/parameter.py:
```diff
@@ -25,16 +25,216 @@
+        pass
+
+
+class ParamDict(StructuredParameter):
```
Oh, you are right. `instantiated` structured params are dicts. Forget about it.

hadware commented:
Bumping this PR... I think it's ready for merge, right?

hbredin commented:
The more I look at this (quite large) PR, the more I think [making ParamDict inherit from Pipeline](https://github.com/pyannote/pyannote-pipeline/pull/34#issuecomment-933807829) is the way to go, as everything is already taken care of in Pipeline.

![image](https://user-images.githubusercontent.com/1080837/149220702-b58bdcb3-0a0f-4393-9f94-0c68a36086a2.png)

hadware commented:
Trying to beat me at my own game, I see...
![image](https://user-images.githubusercontent.com/5477031/149368731-8a6d2249-1763-4b6a-9c05-d5558e3160e9.png)

hadware commented:
But in all honesty I haven't much to say against the idea of making things a bit less redundant between `Pipeline` and `ParamDict`, other than instead of having `ParamDict` inherit from `Pipeline`, i'd rather have some sort of root class that's shared between all parameters (structured or not) and the `Pipeline` class.

If you think that's not how it should be, and just want a quick fix with the  "`ParamDict` inherit from `Pipeline`" solution, I'm okay with it as well. I haven't figured out how this will play out for `ParamList` though.

hbredin commented:
I went ahead and created a separate PR for pipeline-based ParamDict.
I will know update your PR to only keep tests (as those were missing anyway).

hbredin commented:
For some reason, I did not manage to update your PR directly.
Here is the new one with only typing and tests: https://github.com/pyannote/pyannote-pipeline/pull/41

Would you mind checking it that looks good to you?

hadware commented:
I will! Should i close this PR?

hbredin commented:
#41 is now merged. Closing.
